File: retrocpu_boot_monitor/mn1613/test/handshake/handshake_common_test.ts
/**
 * handshake_common.asm（線制御: initiate / send / recv / finalize）
 * 根拠: HandShake.mdc / boot_monitor.mdc / test_framework.mdc
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const HSHK_OK = 0x00;
const HSHK_NG = 0x01;

const BASE_REGS = {
  R2: 0x2222,
  R3: 0x3333,
  R4: 0x4444,
} as const;

const SAVED = { R3: 0x3333, R4: 0x4444 } as const;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ ioMock handshake で 1 ケースを実行する。
 * @param fn 本体（session / mock 利用可）
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * IO が HSHK_REQ_1 を上げるまで待つ。
 * @param mock IO モック
 * @param timeoutMs 上限 ms
 */
async function waitReq1(
  mock: IoBoardHandshakeMock,
  timeoutMs = 2000,
): Promise<void> {
  const t0 = Date.now();
  while (mock.bus.HSHK_REQ_1 !== 1) {
    if (Date.now() - t0 > timeoutMs) {
      throw new Error("timeout waiting HSHK_REQ_1");
    }
    await new Promise((r) => setTimeout(r, 1));
  }
}

/**
 * CPU→IO を initiate / send_byte×N / finalize で送り、IO が同じバイトを受け取る。
 * @param mock IO モック
 * @param bytes 送信バイト（各 0–255）
 * @returns IO が受信した列
 */
async function cpuToIoBytes(
  mock: IoBoardHandshakeMock,
  bytes: number[],
): Promise<Uint8Array> {
  const ioP = mock.io.receive(bytes.length);
  const cpuP = (async () => {
    const init = await session.call("g_hshk_initiate_send", {
      registers: { ...BASE_REGS },
    });
    expect(init.registers.R[0]).toBe(HSHK_OK);
    for (const b of bytes) {
      const sent = await session.call("g_hshk_send_byte", {
      registers: { ...BASE_REGS, R0: b & 0xff },
      });
      expect(sent.registers.R[0]).toBe(HSHK_OK);
    }
    const fin = await session.call("g_hshk_finalize_send", {
      registers: { ...BASE_REGS },
    });
    expect(fin.registers.R[0]).toBe(HSHK_OK);
  })();
  const [received] = await Promise.all([ioP, cpuP]);
  return received;
}

/**
 * IO→CPU を送り、wait_req1_1 / accept / recv_byte×N / finalize で受け取る。
 * @param mock IO モック
 * @param bytes IO が送るバイト
 * @returns recv_byte が R1 に返した受信バイト列
 */
async function ioToCpuBytes(
  mock: IoBoardHandshakeMock,
  bytes: number[],
): Promise<number[]> {
  const ioP = mock.io.send(Uint8Array.from(bytes));
  await waitReq1(mock);
  const wait = await session.call("g_hshk_wait_req1_1", {
      registers: { ...BASE_REGS },
  });
  expect(wait.registers.R[0]).toBe(HSHK_OK);
  const acc = await session.call("g_hshk_accept_request", {
      registers: { ...BASE_REGS },
  });
  expect(acc.registers.R[0]).toBe(HSHK_OK);
  const got: number[] = [];
  for (let i = 0; i < bytes.length; i += 1) {
    const rec = await session.call("g_hshk_recv_byte", {
      registers: { ...BASE_REGS },
    });
    expect(rec.registers.R[0]).toBe(HSHK_OK);
    got.push(rec.registers.R[1] & 0xff);
  }
  const fin = await session.call("g_hshk_finalize_recv", {
      registers: { ...BASE_REGS },
  });
  expect(fin.registers.R[0]).toBe(HSHK_OK);
  await ioP;
  return got;
}

test("CPU→IO 1 バイトが initiate / send / finalize で届く", async () => {
  await withCase(async (s, mock) => {
    const received = await cpuToIoBytes(mock, [0xa5]);
    expect(Array.from(received)).toEqual([0xa5]);
    expect(mock.bus.HSHK_ENA).toBe(0);
    expect(mock.bus.HSHK_REQ_0).toBe(0);
    s.expectRegisters({ R0: HSHK_OK, ...SAVED });
  });
});

test("CPU→IO 複数バイトが順に届く", async () => {
  await withCase(async (s, mock) => {
    const received = await cpuToIoBytes(mock, [0x19, 0x01, 0x64]);
    expect(Array.from(received)).toEqual([0x19, 0x01, 0x64]);
    s.expectRegisters({ R0: HSHK_OK, ...SAVED });
  });
});

test("g_hshk_reg_send16 は 16bit をビッグエンディアン 2 バイトで送る", async () => {
  await withCase(async (s, mock) => {
    const ioP = mock.io.receive(2);
    const cpuP = (async () => {
      const init = await s.call("g_hshk_initiate_send", {
      registers: { ...BASE_REGS },
      });
      expect(init.registers.R[0]).toBe(HSHK_OK);
      const sent = await s.call("g_hshk_reg_send16", {
      registers: { ...BASE_REGS, R0: 0xabcd },
      });
      expect(sent.registers.R[0]).toBe(HSHK_OK);
      const fin = await s.call("g_hshk_finalize_send", {
      registers: { ...BASE_REGS },
      });
      expect(fin.registers.R[0]).toBe(HSHK_OK);
    })();
    const [received] = await Promise.all([ioP, cpuP]);
    expect(Array.from(received)).toEqual([0xab, 0xcd]);
    s.expectRegisters({ R0: HSHK_OK, ...SAVED });
  });
});

test("ENA が 1 のままなら g_hshk_initiate_send は NG", async () => {
  await withCase(async (s, mock) => {
    mock.bus.HSHK_ENA = 1;
    const r = await s.call("g_hshk_initiate_send", {
      registers: { ...BASE_REGS },
    });
    expect(r.registers.R[0]).toBe(HSHK_NG);
    s.expectRegisters({ R3: 0x3333, R4: 0x4444 });
  });
});

test("相手がいない g_hshk_send_byte は DACK 待ちで NG", async () => {
  await withCase(async (s) => {
    const r = await s.call("g_hshk_send_byte", {
      registers: { ...BASE_REGS, R0: 0x5a },
    });
    expect(r.registers.R[0]).toBe(HSHK_NG);
    s.expectRegisters({ R3: 0x3333, R4: 0x4444 });
  });
});

test("IO→CPU 1 バイトが accept / recv / finalize で R1 に入る", async () => {
  await withCase(async (s, mock) => {
    const got = await ioToCpuBytes(mock, [0xc3]);
    expect(got).toEqual([0xc3]);
    expect(mock.bus.HSHK_ENA).toBe(0);
    expect(mock.bus.HSHK_REQ_1).toBe(0);
    s.expectRegisters({ R0: HSHK_OK, ...SAVED });
  });
});

test("IO→CPU 2 バイトを連続受信する", async () => {
  await withCase(async (s, mock) => {
    const got = await ioToCpuBytes(mock, [0x48, 0xab]);
    expect(got).toEqual([0x48, 0xab]);
    s.expectRegisters({ R0: HSHK_OK, ...SAVED });
  });
});

test("g_hshk_wait_req1_1 は REQ_1=1 なら OK", async () => {
  await withCase(async (s, mock) => {
    mock.bus.HSHK_REQ_1 = 1;
    const r = await s.call("g_hshk_wait_req1_1", {
      registers: { ...BASE_REGS },
    });
    expect(r.registers.R[0]).toBe(HSHK_OK);
    s.expectRegisters({ R3: 0x3333, R4: 0x4444 });
  });
});

test("g_hshk_wait_req1_1 は REQ_1 が来なければ NG", async () => {
  await withCase(async (s, mock) => {
    mock.bus.HSHK_REQ_1 = 0;
    const r = await s.call("g_hshk_wait_req1_1", {
      registers: { ...BASE_REGS },
    });
    expect(r.registers.R[0]).toBe(HSHK_NG);
    s.expectRegisters({ R3: 0x3333, R4: 0x4444 });
  });
});

test("g_hshk_finalize_recv は ENA を落として OK を返す", async () => {
  await withCase(async (s, mock) => {
    mock.bus.HSHK_ENA = 1;
    const r = await s.call("g_hshk_finalize_recv", {
      registers: { ...BASE_REGS },
    });
    expect(r.registers.R[0]).toBe(HSHK_OK);
    expect(mock.bus.HSHK_ENA).toBe(0);
    s.expectRegisters({ R3: 0x3333, R4: 0x4444 });
  });
});

test("g_hshk_wait_ena_delay は乱数種を進め R3–R4 を保つ", async () => {
  await withCase(async (s) => {
    const seedBefore = s.readWord(s.wordAddr("GL_RND_SEED"));
    await s.call("g_hshk_wait_ena_delay", {
      registers: { ...BASE_REGS } });
    const seedAfter = s.readWord(s.wordAddr("GL_RND_SEED"));
    expect(seedAfter === seedBefore).toBe(false);
    expect(seedAfter).toBeGreaterThanOrEqual(1);
    s.expectRegisters({ R3: 0x3333, R4: 0x4444 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_common_test.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

const HSHK_OK: u16 = 0x00;
const HSHK_NG: u16 = 0x01;

fn base_regs() -> CallRegisters {
    CallRegisters {
        r2: Some(0x2222),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn wait_req1(mock: &IoBoardHandshakeMock, timeout: Duration) -> Result<(), FrameworkError> {
    let start = Instant::now();
    loop {
        if mock.wires.lock().expect("wires lock").hshk_in_req == 1 {
            return Ok(());
        }
        if start.elapsed() > timeout {
            return Err(FrameworkError::invalid_argument(
                "timeout waiting hshk_in_req",
            ));
        }
        std::thread::yield_now();
    }
}

fn collect_cpu_frame(
    mock: &IoBoardHandshakeMock,
    expected_len: usize,
) -> Result<Vec<u8>, FrameworkError> {
    let mut frame = Vec::with_capacity(expected_len);
    let start = Instant::now();
    while frame.len() < expected_len {
        if let Some(chunk) = mock.take_cpu_to_io_frame() {
            frame.extend(chunk);
            continue;
        }
        if start.elapsed() > Duration::from_millis(2000) {
            return Err(FrameworkError::invalid_argument(
                "timeout collecting cpu_to_io frame",
            ));
        }
        std::thread::yield_now();
    }
    frame.truncate(expected_len);
    Ok(frame)
}

fn cpu_to_io_bytes(
    s: &mut Mn1613AsmSession,
    mock: &IoBoardHandshakeMock,
    bytes: &[u8],
) -> Result<Vec<u8>, FrameworkError> {
    let init = s.call(
        "g_hshk_initiate_send",
        CallOptions {
            registers: Some(base_regs()),
            ..Default::default()
        },
    )?;
    assert_eq!(init.registers.r[0], HSHK_OK);

    for b in bytes {
        let sent = s.call(
            "g_hshk_send_byte",
            CallOptions {
                registers: Some(CallRegisters {
                    r0: Some(*b as u16),
                    ..base_regs()
                }),
                ..Default::default()
            },
        )?;
        assert_eq!(sent.registers.r[0], HSHK_OK);
    }

    let fin = s.call(
        "g_hshk_finalize_send",
        CallOptions {
            registers: Some(base_regs()),
            ..Default::default()
        },
    )?;
    assert_eq!(fin.registers.r[0], HSHK_OK);

    collect_cpu_frame(mock, bytes.len())
}

fn io_to_cpu_bytes(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    bytes: &[u8],
) -> Result<Vec<u8>, FrameworkError> {
    let io = Arc::clone(mock);
    let payload = bytes.to_vec();
    let feeder = std::thread::spawn(move || {
        let mut poll = || std::thread::yield_now();
        io.exchange_with_cpu(&payload, 0, &mut poll)
    });

    wait_req1(mock, Duration::from_millis(5000))?;

    let wait = s.call(
        "g_hshk_wait_req1_1",
        CallOptions {
            registers: Some(base_regs()),
            ..Default::default()
        },
    )?;
    assert_eq!(wait.registers.r[0], HSHK_OK);

    let acc = s.call(
        "g_hshk_accept_request",
        CallOptions {
            registers: Some(base_regs()),
            ..Default::default()
        },
    )?;
    assert_eq!(acc.registers.r[0], HSHK_OK);

    let mut got = Vec::with_capacity(bytes.len());
    for _ in 0..bytes.len() {
        let rec = s.call(
            "g_hshk_recv_byte",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        assert_eq!(rec.registers.r[0], HSHK_OK);
        got.push((rec.registers.r[1] & 0xff) as u8);
    }

    let fin = s.call(
        "g_hshk_finalize_recv",
        CallOptions {
            registers: Some(base_regs()),
            ..Default::default()
        },
    )?;
    assert_eq!(fin.registers.r[0], HSHK_OK);

    let _ = feeder
        .join()
        .map_err(|_| FrameworkError::invalid_argument("io feeder panicked"))??;

    Ok(got)
}

#[test]
fn cpu_to_io_single_byte_arrives_with_initiate_send_finalize() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let received = cpu_to_io_bytes(s, &mock, &[0xa5])?;
        assert_eq!(received, vec![0xa5]);
        {
            let w = mock.wires.lock().expect("wires lock");
            assert_eq!(w.hshk_out_dena, 0);
            assert_eq!(w.hshk_in_req, 0);
        }
        s.expect_registers(
            &CallRegisters {
                r0: Some(HSHK_OK),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn cpu_to_io_multiple_bytes_arrive_in_order() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let received = cpu_to_io_bytes(s, &mock, &[0x19, 0x01, 0x64])?;
        assert_eq!(received, vec![0x19, 0x01, 0x64]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(HSHK_OK),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn send_word_sends_big_endian_two_bytes() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let init = s.call(
            "g_hshk_initiate_send",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        assert_eq!(init.registers.r[0], HSHK_OK);

        let sent = s.call(
            "g_hshk_send_word",
            CallOptions {
                registers: Some(CallRegisters {
                    r0: Some(0xabcd),
                    ..base_regs()
                }),
                ..Default::default()
            },
        )?;
        assert_eq!(sent.registers.r[0], HSHK_OK);

        let fin = s.call(
            "g_hshk_finalize_send",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        assert_eq!(fin.registers.r[0], HSHK_OK);

        let received = collect_cpu_frame(&mock, 2)?;
        assert_eq!(received, vec![0xab, 0xcd]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(HSHK_OK),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn initiate_send_returns_ng_when_ena_is_already_1() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        mock.wires.lock().expect("wires lock").hshk_out_dena = 1;
        let r = s.call(
            "g_hshk_initiate_send",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        assert_eq!(r.registers.r[0], HSHK_NG);
        s.expect_registers(
            &CallRegisters {
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn send_byte_without_partner_returns_ng() -> Result<(), FrameworkError> {
    with_case(|s, _mock| {
        let r = s.call(
            "g_hshk_send_byte",
            CallOptions {
                registers: Some(CallRegisters {
                    r0: Some(0x005a),
                    ..base_regs()
                }),
                ..Default::default()
            },
        )?;
        assert_eq!(r.registers.r[0], HSHK_NG);
        s.expect_registers(
            &CallRegisters {
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn io_to_cpu_single_byte_is_received_in_r1() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let got = io_to_cpu_bytes(s, &mock, &[0xc3])?;
        assert_eq!(got, vec![0xc3]);
        {
            let w = mock.wires.lock().expect("wires lock");
            assert_eq!(w.hshk_out_dena, 0);
            assert_eq!(w.hshk_in_req, 0);
        }
        s.expect_registers(
            &CallRegisters {
                r0: Some(HSHK_OK),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn io_to_cpu_two_bytes_are_received_in_order() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let got = io_to_cpu_bytes(s, &mock, &[0x48, 0xab])?;
        assert_eq!(got, vec![0x48, 0xab]);
        s.expect_registers(
            &CallRegisters {
                r0: Some(HSHK_OK),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn wait_req1_returns_ok_when_req_is_1() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        mock.wires.lock().expect("wires lock").hshk_in_req = 1;
        let r = s.call(
            "g_hshk_wait_req1_1",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        assert_eq!(r.registers.r[0], HSHK_OK);
        s.expect_registers(
            &CallRegisters {
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn wait_req1_returns_ng_when_req_does_not_come() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        mock.wires.lock().expect("wires lock").hshk_in_req = 0;
        let r = s.call(
            "g_hshk_wait_req1_1",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        assert_eq!(r.registers.r[0], HSHK_NG);
        s.expect_registers(
            &CallRegisters {
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn finalize_recv_clears_ena_and_returns_ok() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        mock.wires.lock().expect("wires lock").hshk_out_dena = 1;
        let r = s.call(
            "g_hshk_finalize_recv",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        assert_eq!(r.registers.r[0], HSHK_OK);
        assert_eq!(mock.wires.lock().expect("wires lock").hshk_out_dena, 0);
        s.expect_registers(
            &CallRegisters {
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}

#[test]
fn wait_ena_delay_advances_seed_and_preserves_r3_r4() -> Result<(), FrameworkError> {
    with_case(|s, _mock| {
        let seed_addr = s.word_addr("GL_RND_SEED")?;
        let seed_before = s.read_word(seed_addr);
        let _ = s.call(
            "g_hshk_wait_ena_delay",
            CallOptions {
                registers: Some(base_regs()),
                ..Default::default()
            },
        )?;
        let seed_after = s.read_word(seed_addr);
        assert_ne!(seed_after, seed_before);
        assert!(seed_after >= 1);
        s.expect_registers(
            &CallRegisters {
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
